src/runtime_src/core/edge/drm/zocl/zocl_bo.c

```diff
@@ -662,10 +744,112 @@ void zocl_free_host_bo(struct drm_gem_object *gem_obj)
  * allocating 'count' BOs with total size 'size'; If count < 0, we are
  * freeing 'count' BOs with total size 'size'.
  */
-void zocl_update_mem_stat(struct drm_zocl_dev *zdev, u64 size, int count)
+void zocl_update_mem_stat(struct drm_zocl_dev *zdev, u64 size, int count,
+		uint32_t bank)
 {
+	int i, update_bank = zdev->num_mem;
+
+	/*
+	 * If the 'bank' passed in is a valid bank and its type is
+	 * PL-DDR, we update that bank usage. Otherwise, we go
+	 * through our bank list and find the CMA bank to update
+	 * its usage.
+	 */
+	if (bank < zdev->num_mem &&
+	    zdev->mem[bank].zm_type == ZOCL_MEM_TYPE_PLDDR) {
+		update_bank = bank;
+	} else {
+		for (i = 0; i < zdev->num_mem; i++) {
+			if (zdev->mem[i].zm_used &&
+			    zdev->mem[i].zm_type == ZOCL_MEM_TYPE_CMA) {
+				update_bank = i;
+				break;
+			}
+		}
+	}
+
+	if (update_bank == zdev->num_mem)
+		return;
+
 	write_lock(&zdev->attr_rwlock);
-	zdev->mm_usage.memory_usage += (count > 0) ?  size : -size;
-	zdev->mm_usage.bo_count += count;
+	zdev->mem[update_bank].zm_stat.memory_usage +=
+	    (count > 0) ?  size : -size;
+	zdev->mem[update_bank].zm_stat.bo_count += count;
 	write_unlock(&zdev->attr_rwlock);
 }
+
+/*
+ * Initialize the memory structure in zocl driver based on the memory
+ * topology extracted from xclbin.
+ *
+ * Currently, we could have multiple memory sections but only two type
+ * of them could be marked as used. We identify the memory type by its
+ * tag. If the tag field contains "MIG", it is PL-DDR. Other tags
+ * e.g. "HP", "HPC", it is CMA memory.
+ *
+ * PL-DDR is managed by DRM MM Range Allocator;
+ * CMA is managed by DRM CMA Allocator.
+ */
+void zocl_init_mem(struct drm_zocl_dev *zdev, struct mem_topology *mtopo)
+{
+	struct zocl_mem *memp;
+	int i;
+
+	zdev->num_mem = mtopo->m_count;
+	zdev->mem = vzalloc(zdev->num_mem * sizeof(struct zocl_mem));
+
+	for (i = 0; i < zdev->num_mem; i++) {
+		struct mem_data *md = &mtopo->m_mem_data[i];
+
+		if (!md->m_used)
+			continue;
+
+		memp = &zdev->mem[i];
+		if (md->m_type == MEM_STREAMING) {
+			memp->zm_type = ZOCL_MEM_TYPE_STREAMING;
+			continue;
+		}
+
+		memp->zm_base_addr = md->m_base_address;
+		/* In mem_topology, size is in KB */
+		memp->zm_size = md->m_size * 1024;
+		memp->zm_used = 1;
+
+		if (!strstr(md->m_tag, "MIG")) {
+			memp->zm_type = ZOCL_MEM_TYPE_CMA;
+			continue;
+		}
+
+		memp->zm_mm = vzalloc(sizeof(struct drm_mm));
+		memp->zm_type = ZOCL_MEM_TYPE_PLDDR;
+
+		drm_mm_init(memp->zm_mm, memp->zm_base_addr, memp->zm_size);
+	}
+}
+
+void zocl_clear_mem(struct drm_zocl_dev *zdev)
+{
+	int i;
+
+	mutex_lock(&zdev->mm_lock);
+
+	if (!zdev->mem) {
```

Comment:
We should move the if check up and simply return if mem is NULL. We would not need to lock first and then unlock the mutex if mem is NULL.

Reply:
Fixed. Thanks!